leshan-server-cf/src/main/java/org/eclipse/leshan/server/californium/impl/CaliforniumLwM2mRequestSender.java

```diff
@@ -47,6 +51,7 @@
     private final Set<Endpoint> endpoints;
     private final ObservationRegistry observationRegistry;
     private final LwM2mModelProvider modelProvider;
+    private final Map<String/* registration id */, ArrayList<Request> /* pending coap Request */> pendingRequests = new HashMap<>();
```

Comment:
The `clientsByEp` is a map of the in memory `ClientRegistry` implementation. So you can not consider that this class will be always here. We have a [redis version](https://github.com/eclipse/leshan/blob/redis/leshan-server-demo/src/main/java/org/eclipse/leshan/server/demo/cluster/RedisClientRegistry.java#L50) in development  where this map does not exists.
For me `ClientRegistry` should not handle pending request this is not its responsibility.